pkg/kv/kvserver/raftlog/entry.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+	"go.etcd.io/etcd/raft/v3/raftpb"
+)
+
+// Entry contains data related to a raft log entry. This is the raftpb.Entry
+// itself but also all encapsulated data relevant for command application.
+type Entry struct {
+	Meta enginepb.MVCCMetadata // only set after LoadFromRawValue
+	Ent  raftpb.Entry          // must be pre-populated when calling Load
+	// These fields will all be populated by Load.
+	ID  kvserverbase.CmdIDKey         // always set
+	CC1 *raftpb.ConfChange            // only set for config change
+	CC2 *raftpb.ConfChangeV2          // only set for config change
+	CCC *kvserverpb.ConfChangeContext // only set for config change
+	Cmd kvserverpb.RaftCommand        // always set
+}
+
+// LoadFromRawValue populates the Entry from a raw value, i.e. from bytes as
+// stored on a storage.Engine.
+func (e *Entry) LoadFromRawValue(b []byte) error {
+	if err := protoutil.Unmarshal(b, &e.Meta); err != nil {
+		return errors.Wrap(err, "decoding raft log MVCCMetadata")
+	}
+
+	if err := storage.MakeValue(e.Meta).GetProto(&e.Ent); err != nil {
+		return errors.Wrap(err, "unmarshalling raft Entry")
+	}
+
+	return e.load()
+}
+
+// Load populates the Entry from the Ent field. Will reset the Meta field.
+func (e *Entry) Load() error {
+	e.Meta = enginepb.MVCCMetadata{}
+	return e.load()
+}
+
+func (e *Entry) load() error {
+	if len(e.Ent.Data) == 0 {
+		// Raft-proposed empty entry.
+		return nil
+	}
+
+	var payload []byte
+	switch e.Ent.Type {
+	case raftpb.EntryNormal:
+		e.ID, payload = kvserverbase.DecodeRaftCommand(e.Ent.Data)
+	case raftpb.EntryConfChange:
+		e.CC1 = &raftpb.ConfChange{}
+		if err := protoutil.Unmarshal(e.Ent.Data, e.CC1); err != nil {
+			return errors.Wrap(err, "unmarshalling ConfChange")
+		}
+		e.CCC = &kvserverpb.ConfChangeContext{}
+		if err := protoutil.Unmarshal(e.CC1.Context, e.CCC); err != nil {
+			return errors.Wrap(err, "unmarshalling ConfChangeContext")
+		}
+		payload = e.CCC.Payload
+		e.ID = kvserverbase.CmdIDKey(e.CCC.CommandID)
+	case raftpb.EntryConfChangeV2:
+		e.CC2 = &raftpb.ConfChangeV2{}
+		if err := protoutil.Unmarshal(e.Ent.Data, e.CC2); err != nil {
+			return errors.Wrap(err, "unmarshalling ConfChangeV2")
+		}
+		e.CCC = &kvserverpb.ConfChangeContext{}
+		if err := protoutil.Unmarshal(e.CC2.Context, e.CCC); err != nil {
+			return errors.Wrap(err, "unmarshalling ConfChangeContext")
+		}
+		payload = e.CCC.Payload
+		e.ID = kvserverbase.CmdIDKey(e.CCC.CommandID)
+	default:
+		return errors.AssertionFailedf("unknown entry type %d", e.Ent.Type)
+	}
+
+	// TODO(tbg): can len(payload)==0 if we propose an empty command to wake up leader?
+	// If so, is that a problem here?
+	return errors.Wrap(protoutil.Unmarshal(payload, &e.Cmd), "unmarshalling RaftCommand")
+}
+
+// CC returns CC1 or CC2 as an interface, if set. Otherwise, returns nil.
+func (e *Entry) CC() raftpb.ConfChangeI {
```

Comment:
nit: consider typing this out as `ConfChange`, ditto with the fields. It's too terse when encountered in the code.